BUG: Missing explicit libm dependency in meson.build causes undefined symbol errors

### Pandas version checks

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.


### Reproducible Example

```python
# Build pandas from source
pip install -v -v -v --compile --no-deps --no-cache-dir --no-build-isolation --no-binary :all: .

# Check for undefined symbols
find /usr/local/lib/python*/site-packages/pandas -name "*.so" -exec ldd -r -d {} \; 2>&1 | grep -i "undefined symbol"
```

### Issue Description

Several pandas C extension modules use math functions from `libm` (like `floor`, `ceil`, `round`, etc.) but don't explicitly declare the dependency in their `meson.build` files. This causes undefined symbol errors when building from source, particularly in minimal environments or when using certain toolchains.

Running `ldd -r -d` on the built extension modules shows undefined symbols for math functions. For example:

```bash
find /usr/local/lib/python*/site-packages/pandas -name "*.so" -exec ldd -r -d {} \; 2>&1 | grep -i "undefined symbol" || echo "No undefined symbols found"

[./lib/python3.12/site-packages/pandas/_libs/tslibs/conversion.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffffa4c9d000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffffa4a16000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffffa4c50000)
undefined symbol: log10	(./lib/python3.12/site-packages/pandas/_libs/tslibs/conversion.cpython-312-aarch64-linux-gnu.so)

[./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffff8f9bc000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffff8f432000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffff8f96f000)
undefined symbol: sqrt	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: pow	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: fmod	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: sqrtf	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)


[./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffffbd467000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffffbd163000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffffbd41a000)
undefined symbol: fmod	(./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so)
undefined symbol: pow	(./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so)


[./lib/python3.12/site-packages/pandas/_libs/algos.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffffac1a2000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffffabdad000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffffac155000)
undefined symbol: sqrt	(./lib/python3.12/site-packages/pandas/_libs/algos.cpython-312-aarch64-linux-gnu.so)

[./lib/python3.12/site-packages/pandas/_libs/json.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffff80fd0000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffff80d78000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffff80f83000)
undefined symbol: pow	(./lib/python3.12/site-packages/pandas/_libs/json.cpython-312-aarch64-linux-gnu.so)


[./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffff8f9bc000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffff8f432000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffff8f96f000)
undefined symbol: sqrt	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: pow	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: fmod	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)
undefined symbol: sqrtf	(./lib/python3.12/site-packages/pandas/_libs/groupby.cpython-312-aarch64-linux-gnu.so)


[./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so]:
    linux-vdso.so.1 (0x0000ffffbd467000)
    libc.so.6 => /lib64/libc.so.6 (0x0000ffffbd163000)
    /lib/ld-linux-aarch64.so.1 (0x0000ffffbd41a000)
undefined symbol: fmod	(./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so)
undefined symbol: pow	(./lib/python3.12/site-packages/pandas/_libs/sparse.cpython-312-aarch64-linux-gnu.so)
```

This affects at least the following modules:
- `algos`
- `groupby`
- `sparse`
- `json`
- `conversion` (in tslibs)

### Proposed Solution

Add explicit `libm` dependency to the affected extension modules in the meson.build files.

1. Declare the libm dependency:
```meson
cc = meson.get_compiler('c')
m_dep = cc.find_library('m', required: true)
```

2. Add `m_dep` to the dependencies of affected modules:
```meson
'algos': {
    'sources': ['algos.pyx', _algos_common_helper, _algos_take_helper],
    'deps': [_khash_primitive_helper_dep, m_dep],  # Add m_dep
},
'groupby': {'sources': ['groupby.pyx'], 'deps': [m_dep]},
'sparse': {'sources': ['sparse.pyx', _sparse_op_helper], 'deps': [m_dep]},
'json': {
    'sources': [
        'src/vendored/ujson/python/ujson.c',
        'src/vendored/ujson/python/objToJSON.c',
        'src/vendored/ujson/python/JSONtoObj.c',
        'src/vendored/ujson/lib/ultrajsonenc.c',
        'src/vendored/ujson/lib/ultrajsondec.c'
    ],
    'deps': [m_dep]
},
```

Changes needed in `pandas/_libs/tslibs/meson.build`:

1. Declare the libm dependency at the top of the file
2. Add to conversion module:
```meson
'conversion': {'sources': ['conversion.pyx'], 'deps': [m_dep]},
```

### Expected Behavior

All C extensions that use math functions should explicitly link against `libm` to ensure symbols are properly resolved.

As a workaround, patch the necessary changes:

```bash
PANDAS_LIBM_MESON_BUILD_FILE_PATH="./pandas/_libs/meson.build"
PANDAS_LIBM_TSLIBS_MESON_BUILD_FILE_PATH="./pandas/_libs/tslibs/meson.build"

# 1. Add m_dep declaration before subdir('tslibs')
sed -i "/^subdir('tslibs')/i\\
cc = meson.get_compiler('c')\\
m_dep = cc.find_library('m', required: true)\\
" "${PANDAS_LIBM_MESON_BUILD_FILE_PATH}"

# 2. Add m_dep to algos (already has deps, convert to list and add m_dep)
sed -i "/^    'algos':/s/'deps': _khash_primitive_helper_dep/'deps': [_khash_primitive_helper_dep, m_dep]/" "${PANDAS_LIBM_MESON_BUILD_FILE_PATH}"

# 3. Add deps with m_dep to groupby (no existing deps)
sed -i "/^    'groupby':/s/},/, 'deps': [m_dep]},/" "${PANDAS_LIBM_MESON_BUILD_FILE_PATH}"

# 4. Add deps with m_dep to sparse (no existing deps)
sed -i "/^    'sparse':/s/},/, 'deps': [m_dep]},/" "${PANDAS_LIBM_MESON_BUILD_FILE_PATH}"

# 5. Add deps with m_dep to json (multi-line entry)
sed -i "/^    'json':/,/]},/{s/]},/], 'deps': [m_dep]},/}" "${PANDAS_LIBM_MESON_BUILD_FILE_PATH}"

# 6. Add m_dep declaration to tslibs/meson.build (at start of file)
sed -i "1i\\
cc = meson.get_compiler('c')\\
m_dep = cc.find_library('m', required: true)\\
" "${PANDAS_LIBM_TSLIBS_MESON_BUILD_FILE_PATH}"

# 7. Add deps with m_dep to conversion in tslibs
sed -i "/^    'conversion':/s/},/, 'deps': [m_dep]},/" "${PANDAS_LIBM_TSLIBS_MESON_BUILD_FILE_PATH}"
```

### Installed Versions

INSTALLED VERSIONS
------------------
commit                : 9c8bc3e55188c8aff37207a74f1dd144980b8874
python                : 3.12.7
python-bits           : 64
OS                    : Linux
OS-release            : 6.10.14-linuxkit
Version               : #1 SMP Wed Sep 10 06:47:45 UTC 2025
machine               : aarch64
processor             : aarch64
byteorder             : little
LC_ALL                : en_US.UTF-8
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.3.3
numpy                 : 2.2.0.dev0+git20241122.b47bd25
pytz                  : 2024.2
dateutil              : 2.9.0.post0
pip                   : 25.3
Cython                : 3.0.11
sphinx                : None
IPython               : 8.29.0
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
blosc                 : None
bottleneck            : 1.4.2
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2024.10.0
html5lib              : 1.1
hypothesis            : 6.119.4
gcsfs                 : 2024.10.0
jinja2                : 3.1.4
lxml.etree            : 5.3.0
matplotlib            : 3.9.2
numba                 : None
numexpr               : 2.10.1
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
psycopg2              : 2.9.10
pymysql               : None
pyarrow               : None
pyreadstat            : None
pytest                : None
python-calamine       : None
pyxlsb                : None
s3fs                  : 2024.10.0
scipy                 : 1.15.0.dev0+git20241122.c75f588
sqlalchemy            : 2.0.36
tables                : None
tabulate              : 0.9.0
xarray                : 2024.10.0
xlrd                  : None
xlsxwriter            : None
zstandard             : 0.23.0
tzdata                : 2024.2
qtpy                  : None
pyqt5                 : None
>>>


### Environment

- **pandas version**: 2.3.3
- **Build system**: meson
- **Platform**: Amz Linux 2023 aarch64
- **Python version**: 3.12.7

--- a/pandas/_libs/meson.build
+++ b/pandas/_libs/meson.build
@@ -50,7 +50,6 @@ _khash_primitive_helper_dep = declare_dependency(
     sources: _khash_primitive_helper,
 )
 
-cc = meson.get_compiler('c')
 m_dep = cc.find_library('m', required: false)
 
 cdata = configuration_data()

--- a/pandas/_libs/meson.build
+++ b/pandas/_libs/meson.build
@@ -50,6 +50,9 @@ _khash_primitive_helper_dep = declare_dependency(
     sources: _khash_primitive_helper,
 )
 
+cc = meson.get_compiler('c')
+m_dep = cc.find_library('m', required: false)
+
 cdata = configuration_data()
 if cy.version().version_compare('>=3.1.0')
     cdata.set('freethreading_compatible', '1')
@@ -70,10 +73,10 @@ libs_sources = {
     # numpy include dir is implicitly included
     'algos': {
         'sources': ['algos.pyx', _algos_common_helper, _algos_take_helper],
-        'deps': _khash_primitive_helper_dep,
+        'deps': [_khash_primitive_helper_dep, m_dep],
     },
     'arrays': {'sources': ['arrays.pyx']},
-    'groupby': {'sources': ['groupby.pyx']},
+    'groupby': {'sources': ['groupby.pyx'], 'deps': [m_dep]},
     'hashing': {'sources': ['hashing.pyx']},
     'hashtable': {
         'sources': [
@@ -126,6 +129,7 @@ libs_sources = {
             'src/vendored/ujson/lib/ultrajsonenc.c',
             'src/vendored/ujson/lib/ultrajsondec.c',
         ],
+        'deps': [m_dep],
     },
     'ops': {'sources': ['ops.pyx']},
     'ops_dispatch': {'sources': ['ops_dispatch.pyx']},
@@ -133,7 +137,7 @@ libs_sources = {
     'reshape': {'sources': ['reshape.pyx']},
     'sas': {'sources': ['sas.pyx']},
     'byteswap': {'sources': ['byteswap.pyx']},
-    'sparse': {'sources': ['sparse.pyx', _sparse_op_helper]},
+    'sparse': {'sources': ['sparse.pyx', _sparse_op_helper], 'deps': [m_dep]},
     'tslib': {'sources': ['tslib.pyx']},
     'testing': {'sources': ['testing.pyx']},
     'writers': {'sources': ['writers.pyx']},

--- a/pandas/_libs/tslibs/meson.build
+++ b/pandas/_libs/tslibs/meson.build
@@ -4,7 +4,7 @@ tslibs_sources = {
     'base': {'sources': ['base.pyx']},
     'ccalendar': {'sources': ['ccalendar.pyx']},
     'dtypes': {'sources': ['dtypes.pyx']},
-    'conversion': {'sources': ['conversion.pyx']},
+    'conversion': {'sources': ['conversion.pyx'], 'deps': [m_dep]},
     'fields': {'sources': ['fields.pyx']},
     'nattype': {'sources': ['nattype.pyx']},
     'np_datetime': {'sources': ['np_datetime.pyx']},

